src/main/java/org/extism/sdk/HostFunction.java

```diff
@@ -1,69 +1,72 @@
+    public HostFunction(String name, LibExtism.ExtismValType[] params, LibExtism.ExtismValType[] returns, ExtismFunction<T> func, T userData) {
 
         this.name = name;
         this.params = params;
         this.returns = returns;
         this.userData = userData;
-        this.callback = (Pointer currentPlugin,
-                         LibExtism.ExtismVal inputs,
-                         int nInputs,
-                         LibExtism.ExtismVal outs,
-                         int nOutputs,
-                         Pointer data) -> {
-
-            LibExtism.ExtismVal[] outputs = (LibExtism.ExtismVal []) outs.toArray(nOutputs);
-
-            f.invoke(
-                    new ExtismCurrentPlugin(currentPlugin),
-                    (LibExtism.ExtismVal []) inputs.toArray(nInputs),
-                    outputs,
-                    userData
-            );
-
-            for (LibExtism.ExtismVal output : outputs) {
-                convertOutput(output, output);
-            }
-        };
+
+        int[] inputTypeValues = Arrays.stream(this.params).mapToInt(r -> r.v).toArray();
+        int[] outputTypeValues = Arrays.stream(this.returns).mapToInt(r -> r.v).toArray();
+        InternalExtismFunction callback = createCallbackFunction(func, userData);
+        Pointer userDataPointer = userData != null ? userData.getPointer() : null;
 
         this.pointer = LibExtism.INSTANCE.extism_function_new(
-                this.name,
-                Arrays.stream(this.params).mapToInt(r -> r.v).toArray(),
-                this.params.length,
-                Arrays.stream(this.returns).mapToInt(r -> r.v).toArray(),
-                this.returns.length,
-                this.callback,
-                userData.map(PointerType::getPointer).orElse(null),
+                name,
+                inputTypeValues,
+                inputTypeValues.length,
+                outputTypeValues,
+                outputTypeValues.length,
+                callback,
+                userDataPointer,
                 null
         );
     }
 
-    void convertOutput(LibExtism.ExtismVal original, LibExtism.ExtismVal fromHostFunction) {
-        if (fromHostFunction.t != original.t)
-            throw new ExtismException(String.format("Output type mismatch, got %d but expected %d", fromHostFunction.t, original.t));
+    private InternalExtismFunction createCallbackFunction(ExtismFunction<T> func, T userData) {
+        return (Pointer pluginPointer, ExtismVal inputs, int nInputs, ExtismVal outputs, int nOutputs, Pointer data) -> {
+
+            var outputValues = (ExtismVal[]) outputs.toArray(nOutputs);
+            var inputValues = (ExtismVal[]) inputs.toArray(nInputs);
+            var currentPlugin = new ExtismCurrentPlugin(pluginPointer);
+
+            func.invoke(currentPlugin, inputValues, outputValues, Optional.ofNullable(userData));
+
+            for (ExtismVal output : outputValues) {
+                convertOutput(output, output);
```

Comment:
This looks wrong, but is like that also in the current implementation here: https://github.com/extism/java-sdk/blob/main/src/main/java/org/extism/sdk/HostFunction.java#L46

Reply:
Hmm yeah seems like a mistake for these to both be `output`

Reply:
I think the original convertOutput method, called via `convertOutput(output, output);`

```
    void convertOutput(LibExtism.ExtismVal original, LibExtism.ExtismVal fromHostFunction) {
        if (fromHostFunction.t != original.t)
            throw new ExtismException(String.format("Output type mismatch, got %d but expected %d", fromHostFunction.t, original.t));

        if (fromHostFunction.t == LibExtism.ExtismValType.I32.v) {
            original.v.setType(Integer.TYPE);
            original.v.i32 = fromHostFunction.v.i32;
        } else if (fromHostFunction.t == LibExtism.ExtismValType.I64.v) {
            original.v.setType(Long.TYPE);
            original.v.i64 = fromHostFunction.v.i64;
        } else if (fromHostFunction.t == LibExtism.ExtismValType.F32.v) {
            original.v.setType(Float.TYPE);
            original.v.f32 = fromHostFunction.v.f32;
        } else if (fromHostFunction.t == LibExtism.ExtismValType.F64.v) {
            original.v.setType(Double.TYPE);
            original.v.f64 = fromHostFunction.v.f64;
        } else
            throw new ExtismException(String.format("Unsupported return type: %s", original.t));
    }
```

can be converted to:
```java
    private void coerceType(ExtismVal value) {

        if (value.t == LibExtism.ExtismValType.I32.v) {
            value.v.setType(Integer.TYPE);
        } else if (value.t == LibExtism.ExtismValType.I64.v) {
            value.v.setType(Long.TYPE);
        } else if (value.t == LibExtism.ExtismValType.F32.v) {
            value.v.setType(Float.TYPE);
        } else if (value.t == LibExtism.ExtismValType.F64.v) {
            value.v.setType(Double.TYPE);
        } else {
            throw new ExtismException(String.format("Unsupported return type: %s", value.t));
        }
    }
```
with call `coerceType(output);`.

Reply:
Yeah makes sense. yeah the setType seems to be the only real mutation!

Reply:
Do we need to do the same with the inputValues?

Reply:
just adapted the code for the type coercion of the output parameters.